src/error-snippets/utils.ts

```diff
@@ -0,0 +1,85 @@
+import path from "path";
+import { fileURLToPath } from "url";
+import fs from "fs-extra";
+import { codeFrameColumns } from "@babel/code-frame";
+import { getErrorTitle } from "../browser/stacktrace/utils";
+import { SNIPPET_LINES_ABOVE, SNIPPET_LINES_BELOW, SOURCE_MAP_URL_COMMENT } from "./constants";
+
+interface FormatFileNameHeaderOpts {
+    line: number;
+    linesAbove: number;
+    linesBelow: number;
+}
+
+interface FormatErrorSnippetOpts {
+    file: string;
+    source: string;
+    location: { line: number; column: number };
+}
+
+export const softFileURLToPath = (fileName: string): string => {
+    if (!fileName.startsWith("file://")) {
+        return fileName;
+    }
+
+    try {
+        return fileURLToPath(fileName);
+    } catch (_) {
+        return fileName;
+    }
+};
+
+export const formatFileNameHeader = (fileName: string, opts: FormatFileNameHeaderOpts): string => {
+    const lineNumberWidth = String(opts.line - opts.linesAbove).length;
+    const offsetWidth = String(opts.line + opts.linesBelow).length;
+
+    const filePath = softFileURLToPath(fileName);
+    const relativeFileName = path.isAbsolute(filePath) ? path.relative(process.cwd(), filePath) : filePath;
+    const lineNumberOffset = ".".repeat(lineNumberWidth).padStart(offsetWidth);
+    const offset = `  ${lineNumberOffset} |`;
+    const grayBegin = "\x1B[90m";
+    const grayEnd = "\x1B[39m";
+
+    return [offset + ` // ${relativeFileName}`, offset].map(line => grayBegin + line + grayEnd + "\n").join("");
```

Comment:
do I understand correctly that you are joing strings in this format in order to avoid gluing adjacent strings?

Reply:
Yes.

`offset` is `   .. | ` part. So it doesn't look like this:
```
// file-name.ts

   17 |
 > 18 |         const elem = await browser.$('not-existing').waitForDisplayed();
      |                     ^ Error: element ("not-existing") still not displayed after 5000ms
```
but this:
```
   .. | // file-name.ts
   .. |
   17 |
 > 18 |         const elem = await browser.$('not-existing').waitForDisplayed();
      |                     ^ Error: element ("not-existing") still not displayed after 5000ms
```
Then we paint each line with `gray` (but each line separately, so our datetime prefix won't be painted in gray)